lib/src/message.dart

```diff
@@ -9,22 +9,38 @@ import 'package:collection/collection.dart';
 import 'package:http_parser/http_parser.dart';
 
 import 'body.dart';
+import 'headers.dart';
 import 'shelf_unmodifiable_map.dart';
 import 'util.dart';
 
 Body extractBody(Message message) => message._body;
 
 /// The default set of headers for a message created with no body and no
 /// explicit headers.
-final _defaultHeaders =
-    ShelfUnmodifiableMap<String>({'content-length': '0'}, ignoreKeyCase: true);
+final _defaultHeaders = Headers.from({
+  'content-length': ['0'],
+});
 
 /// Represents logic shared between [Request] and [Response].
 abstract class Message {
+  final Headers _headers;
+
   /// The HTTP headers.
   ///
-  /// The value is immutable.
-  final Map<String, String> headers;
+  /// If a header occurs more than once in the query string, they are mapped to
+  /// by concatenating them with a comma.
+  ///
+  /// The returned map is unmodifiable.
+  Map<String, String> get headers => _headers.singleValues;
+
+  /// The HTTP headers with multiple values.
+  ///
+  /// If a header occurs only once, its value is a singleton list.
+  /// If a header occurs with no value, the empty string is used as the value
+  /// for that occurrence.
+  ///
+  /// The returned map and the lists it contains are unmodifiable.
```

Comment:
Unrelated to this PR, but it would be really lovely if the documentation said that headers were case-insensitive :)

Reply:
Done.